src/configuration/configuration_manager.cpp

```diff
@@ -0,0 +1,149 @@
+}
+void SET_STRING(const std::string &name, const std::string &value) {
+  peloton::configuration::ConfigurationManager::GetInstance()->SetValue<std::string>(name, value);
+}
+
+namespace peloton {
+namespace configuration {
+
+ConfigurationManager* ConfigurationManager::GetInstance() {
+  static std::unique_ptr<ConfigurationManager> config_manager(
+          new ConfigurationManager());
+  return config_manager.get();
+}
+
+template<typename T>
+T ConfigurationManager::GetValue(const std::string &name) {
+  auto param = config.find(name);
+  if (param == config.end()) {
+    throw new Exception("no such configuration: " + name);
+  }
+  return to_value<T>(param->second.value);
+}
+
+template<typename T>
+void ConfigurationManager::SetValue(const std::string &name, const T &value) {
+  auto param = config.find(name);
+  if (param == config.end()) {
+    throw new Exception("no such configuration: " + name);
+  }
+  switch (param->second.value_type) {
+    case type::TypeId::INTEGER:
+      *(uint64_t*)(param->second.value) = *(uint64_t*)(&value);
+      break;
+    case type::TypeId::BOOLEAN:
+      *(bool*)(param->second.value) = *(bool*)(&value);
+      break;
+    case type::TypeId::VARCHAR:
+      *(std::string*)(param->second.value) = *(std::string*)(&value);
+      break;
+    default:
+      throw new Exception("unsupported type");
+  }
+  if (catalog_initialized) {
+    insert_into_catalog(param->first, param->second);
+  }
+}
+
+template<typename T>
+void ConfigurationManager::DefineConfig(const std::string &name, void* value, type::TypeId type,
+                                        const std::string &description, const T &default_value,
+                                        bool is_mutable, bool is_persistent) {
+  if (config.count(name) > 0) {
+    throw Exception("configuration " + name + " already exists");
+  }
+  T tmp(default_value);
+  config[name] = Param(value, description, type, to_string(&tmp, type),
+                       is_mutable, is_persistent);
+  if (catalog_initialized) {
+    insert_into_catalog(name, config[name]);
+  }
+}
+
+void ConfigurationManager::InitializeCatalog() {
+  auto config_catalog = peloton::catalog::ConfigCatalog::GetInstance();
+
+  auto &txn_manager = concurrency::TransactionManagerFactory::GetInstance();
+  auto txn = txn_manager.BeginTransaction();
+  type::AbstractPool *pool = pool_.get();
+
+  for (auto conf : config) {
+    config_catalog->DeleteConfig(conf.first, txn);
+    config_catalog->InsertConfig(conf.first, to_string(conf.second.value, conf.second.value_type),
+                                 conf.second.value_type,
+                                 conf.second.desc, "", "", conf.second.default_value,
+                                 conf.second.is_mutable, conf.second.is_persistent,
+                                 pool, txn);
+  }
+  txn_manager.CommitTransaction(txn);
+  catalog_initialized = true;
+}
+
+void ConfigurationManager::PrintConfiguration() {
```

Comment:
Don't create "print" functions. `ConfigurationManager` should extend `Printable` and you should implement a `GetInfo` method.